From: Takashi Iwai <tiwai@suse.de>
Subject: Fix the sample rates on Revo 7.1
Patch-mainline: not yet
References: 

Fixed the available sample rates on Revo 7.1 board.
AK codecs don't support rates below 32kHz.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/pci/ice1712/ice1712.h linux/sound/pci/ice1712/ice1712.h
--- linux-2.6.11/sound/pci/ice1712/ice1712.h	2005-03-08 17:49:42.000000000 +0100
+++ linux/sound/pci/ice1712/ice1712.h	2005-03-07 15:02:11.000000000 +0100
@@ -336,6 +336,7 @@ struct _snd_ice1712 {
 
 	struct semaphore open_mutex;
 	snd_pcm_substream_t *pcm_reserved[4];
+	snd_pcm_hw_constraint_list_t *hw_rates; /* card-specific rate constraints */
 
 	unsigned int akm_codecs;
 	akm4xxx_t *akm;
diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/pci/ice1712/ice1724.c linux/sound/pci/ice1712/ice1724.c
--- linux-2.6.11/sound/pci/ice1712/ice1724.c	2005-03-08 17:49:42.000000000 +0100
+++ linux/sound/pci/ice1712/ice1724.c	2005-03-07 15:04:25.000000000 +0100
@@ -718,6 +718,13 @@ static snd_pcm_hardware_t snd_vt1724_2ch
 static int set_rate_constraints(ice1712_t *ice, snd_pcm_substream_t *substream)
 {
 	snd_pcm_runtime_t *runtime = substream->runtime;
+	if (ice->hw_rates) {
+		/* hardware specific */
+		runtime->hw.rate_min = ice->hw_rates->list[0];
+		runtime->hw.rate_max = ice->hw_rates->list[ice->hw_rates->count - 1];
+		runtime->hw.rates = SNDRV_PCM_RATE_KNOT;
+		return snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_RATE, ice->hw_rates);
+	}
 	if (ice->eeprom.data[ICE_EEP2_ACLINK] & VT1724_CFG_PRO_I2S) {
 		/* I2S */
 		/* VT1720 doesn't support more than 96kHz */
diff -rupN --exclude-from=/home/iwai/alsa-kernel-excludes linux-2.6.11/sound/pci/ice1712/revo.c linux/sound/pci/ice1712/revo.c
--- linux-2.6.11/sound/pci/ice1712/revo.c	2005-03-08 17:49:42.000000000 +0100
+++ linux/sound/pci/ice1712/revo.c	2005-03-07 15:05:49.000000000 +0100
@@ -128,6 +128,17 @@ static struct snd_ak4xxx_private akm_rev
 	.mask_flags = 0,
 };
 
+static unsigned int rates[] = {
+	32000, 44100, 48000, 64000, 88200, 96000,
+	176400, 192000,
+};
+
+static snd_pcm_hw_constraint_list_t revo_rates = {
+	.count = ARRAY_SIZE(rates),
+	.list = rates,
+	.mask = 0,
+};
+
 static int __devinit revo_init(ice1712_t *ice)
 {
 	akm4xxx_t *ak;
@@ -162,6 +173,8 @@ static int __devinit revo_init(ice1712_t
 		break;
 	}
 
+	ice->hw_rates = &revo_rates; /* AK codecs don't support lower than 32k */
+
 	return 0;
 }
 
